### PR TITLE
Move Filters to webapi

### PR DESCRIPTION
**Pull request by auslin-aot**
# Issue Tracking

JIRA: 
Issue Type: FEATURE
https://aottech.atlassian.net/browse/FWF-1659

# Changes
Move Filters to webapi

- Capturing filter data in webAPI.
- Endpoints to create new filter, get all filters, get filters for specific user, get, update & delete filter by id.
- Deleting filter by id updates filter status to inactive
- Only filter with  status=active will be available for get, update, delete endpoints

# Screenshots
**Create filter:**
image

**List all filters**
image

**List  filters  of a user/role**
image

**Get filter by id**
image

**Update filter by id**
image

**Delete filter by id**
image

### CODE REVIEW

File: forms-flow-api/src/formsflow_api/models/filter.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+from formsflow_api.models.db import db
+
+from .audit_mixin import AuditDateTimeMixin, AuditUserMixin
+
+
+class Filter(AuditDateTimeMixin, AuditUserMixin, BaseModel, db.Model):
+    """This class manages filter information."""
+
+    id = db.Column(db.Integer, primary_key=True)
+    tenant = db.Column(db.String, nullable=True)
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/models/filter.py:20]: Please add an index to tenant column

File: forms-flow-api/src/formsflow_api/resources/filter.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+                "message": "Invalid request data",
+            }, BAD_REQUEST
+            return response, status
+        except Exception as unexpected_error:
+            current_app.logger.warning(unexpected_error)
+            raise unexpected_error
+
+
+@cors_preflight("GET, OPTIONS")
+@API.route("/users", methods=["GET", "OPTIONS"])
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/resources/filter.py:60]: may be singular name makes sense here ? /user

File: forms-flow-api/src/formsflow_api/services/filter.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+filter_schema = FilterSchema()
+
+
+class FilterService:
+    """This class manages filter service."""
+
+    @staticmethod
+    @user_context
+    def get_all_filter(**kwargs):
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/services/filter.py:19]: get_all_filter**s** might be good here.

File: forms-flow-api/src/formsflow_api/services/filter.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        """Create Filter."""
+        user: UserContext = kwargs["user"]
+        filter_payload["tenant"] = user.tenant_key
+        filter_payload["created_by"] = user.user_name
+        filter_data = Filter.create_filter_from_dict(filter_payload)
+        return filter_schema.dump(filter_data)
+
+    @staticmethod
+    @user_context
+    def get_user_filter(**kwargs):
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/services/filter.py:37]: again plural function name would be better

File: forms-flow-api/src/formsflow_api/resources/filter.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
+"""API endpoints for filter resource."""
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/resources/filter.py:1]: Need to secure all these endpoints for `formsflow-reviewer` role.

**Review by sumesh-aot — COMMENTED** [review]
Please add some test cases

File: forms-flow-api/src/formsflow_api/resources/filter.py (hunk 5 above)
Comment (abhilash-aot) [reply forms-flow-api/src/formsflow_api/resources/filter.py:1]: Filter creation access can be granted to a reviewer admin ? or designer ?

Comment (sumesh-aot) [reply forms-flow-api/src/formsflow_api/resources/filter.py:1]: I was thinking of letting the reviewer create the filter right on the tasks page.

**Comment by auslin-aot** [comment]
Added testcases and changes.

File: forms-flow-api/src/formsflow_api/resources/filter.py (hunk 5 above)
Comment (auslin-aot) [reply forms-flow-api/src/formsflow_api/resources/filter.py:1]: > Need to secure all these endpoints for `formsflow-reviewer` role.

done

**Review by SnehaSuresh-aot — APPROVED** [review]
LGTM

**Review by abhilash-aot — APPROVED** [review]

**Comment by abhilash-aot** [comment]
filter changes for task - Task list to support new adding and deleting new filters

File: forms-flow-api/src/formsflow_api/models/filter.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    roles = db.Column(ARRAY(db.String), nullable=True, comment="Applicable roles")
+    users = db.Column(ARRAY(db.String), nullable=True, comment="Applicable users")
+    status = db.Column(db.String(10), nullable=True)
+
+    @classmethod
+    def find_all_active_filters(cls, tenant: str = None) -> List[Filter]:
+        """Find all active filters."""
+        query = cls.query.filter(Filter.status == str(FilterStatus.ACTIVE.value))
+        if tenant:
+            query = cls.query.filter(Filter.tenant == tenant)
```
Comment (sreehari-aot) [inline forms-flow-api/src/formsflow_api/models/filter.py:36]: This should be `query = query.filter(Filter.tenant == tenant)` right ? for multitenant also we are expecting to return active filters for the corresponding tenant right? The current code will overwrite the initial query if `tenant` is true, returning all the tenant-specific filters despite the status.

Comment (auslin-aot) [reply forms-flow-api/src/formsflow_api/models/filter.py:36]: good catch. I missed that. Will update

Comment (auslin-aot) [reply forms-flow-api/src/formsflow_api/models/filter.py]: done

**Review by sreehari-aot — APPROVED** [review]
LGTM!

**Pull request merged**